File: Python/textbook_track/chapter05/ba5a.py
# Find the Minimum Number of Coins Needed to Make Change
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split(',')]


def calc_minimum_number_of_coins(amount, coins):
    changes = [0] * (amount+1)
    for money in range(1, amount+1):
        changes[money] = 1 + min(changes[k] for coin in coins if (k := money - coin) >= 0)
    return changes[amount]


def main():
    data = sys.stdin.read().splitlines()
    money = int(data[0])
    coins = convert_to_intlist(data[1])
    result = calc_minimum_number_of_coins(money, coins)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter05/BA5A.scala
package TextbookTrack.Chapter05

object BA5A {
  private def convertToIntList(line: String): List[Int] = line.split(",").map(_.toInt).toList

  def calcMinimumNumberOfCoins(amount: Int, coins: List[Int]): Int = {
    val changes: Array[Int] = Array.fill(amount + 1)(0)
    (1 to amount).foreach{
      money => changes(money) = 1 + coins.withFilter(money >= _).map(coin => changes(money - coin)).min
    }
    changes(amount)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val money: Int = reader.next().toInt
    val coins: List[Int] = convertToIntList(reader.next())
    val result: Int = calcMinimumNumberOfCoins(money, coins)
    println(result)
  }
}
